### PR TITLE
Allow running multiple Desktop instances on the same machine

### LINKED ISSUE #6213: Run several Status Desktop builds simultaneously
### User Story

It would be nice to be able to run several Jenkins builds simultaneously on the same machine. For instance, this would help QA team a lot.

### Description
A possible scenario is simultaneous usage/testing of stable, nightly and PR builds.

Note that this is different from https://github.com/status-im/status-react/issues/5581, as it implied using dev builds together with Jenkins builds, not 2 Jenkins builds simultaneously.

### Solution
Each instance of Status Desktop has to use unique values for:
  - Node.js port
  - Status node port
  - Status node datadir
  - App bundle name

A possible solution would e.g. involve modifying Jenkins so that a build configuration can be selected from a list of hardcoded options. There must be some better solutions though.

### PR DESCRIPTION
**Pull request by vitvly**
Fixes #6213 

Depends on https://github.com/status-im/react-native-desktop/pull/387

Following changes are introduced:

1. Value of `REACT_SERVER_PORT` env variable is passed via `--port` param to the `ubuntu-server` subprocess that is launched inside a bundle.
2. Values of `STATUS_NODE_PORT`, `STATUS_DATA_DIR` env variables are passed to RN root component via initialProps (
https://facebook.github.io/react-native/docs/communication-ios#passing-properties-from-native-to-react-native) and afterwards used for base node configuration.
3. In order to make item 2 work in dev environment, figwheel bridge had to be modified so that to use React Component as a `root_el`, instead of React Element. This is necessary to pass initialProps to `app-root`, as figwheel wraps `app-root` in its own `reloader`.

**QA notes:**
Set `REACT_SERVER_PORT`, `STATUS_NODE_PORT`, `STATUS_DATA_DIR` env variable values prior to running each instance of Status.app from terminal, e.g.:
```
export REACT_SERVER_PORT=5001
export STATUS_NODE_PORT=10001
export STATUS_DATA_DIR=~/status-data/instance1 # Note the absolute path here
open Status.app
```

### CODE REVIEW

**Comment by ghost** [comment]
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

**Review by maxhora — APPROVED** [review]

File: desktop_files/package.json [hunk 1]
```diff
@@ -55,7 +55,7 @@
     "querystring-es3": "0.2.1",
     "react": "16.4.1",
     "react-dom": "16.4.2",
-    "react-native": "git+https://github.com/status-im/react-native-desktop.git#v0.56.1",
+    "react-native": "git+https://github.com/status-im/react-native-desktop.git#pass-initial-props",
```
Comment (vkjr) [inline desktop_files/package.json]: Just a note to not forget this before commit

**Review by vkjr — APPROVED** [review]

**Review by pedropombeiro — APPROVED** [review]

**Comment by churik** [comment]
@siphiuel 
Thanks a lot for your PR!
It is working, so I can run 2 app instances at the same time - this PR build https://github.com/status-im/status-react/pull/6666#issuecomment-439097837 and current nightly.
Several questions:
- `export STATUS_NODE_DATADIR=/ethereum/mainnet_rpc_test # Note the relative path here`
I suppose that this means that `/ethereum/mainnet_rpc_test` will be created inside STATUS_DB_DIR.
But in fact, it is created on default location on the same level as mainnet_rpc.
dert

- where I can find `StatusIm.log` for the new build?

- would you mind to update docs in the same PR to provide instructions on how to launch several builds?

**Comment by vitvly** [comment]
@Churik thanks for testing this!

1. `STATUS_NODE_DATADIR` - yes, `/ethereum/mainnet_rpc_test` is going to be relative to the default location.
2. `StatusIm.log` - I think both builds write to the same location. Will fix this.

I think i need to improve the UX of this pr, so that instead of setting 2 directories (`datadir` and `db_dir`) we just set one default location that will contain Realm files, Geth data, and logs. Will try to figure this out.

**Comment by churik** [comment]
nice thought! Will waiting for next commit

**Comment by vitvly** [comment]
@churik Fixed and changed QA notes. Also added docs PR: https://github.com/status-im/status.im/pull/154. Now the location for Realm files, logs, and Geth data is specified via single `STATUS_DATA_DIR` environment variable.

**Review by antdanchenko — APPROVED** [review]

**Comment by churik** [comment]
@siphiuel 
1) it is perfectly working on Mac OSx, but on Linux Ubuntu I've got blank screen after launching AppImage.
Steps:
```
parallels@parallels-vm:~/Downloads$ export REACT_SERVER_PORT=5001
parallels@parallels-vm:~/Downloads$ export STATUS_NODE_PORT=10001

parallels@parallels-vm:~/Downloads$ export STATUS_DATA_DIR=/home/parallels/Downloads/st
parallels@parallels-vm:~/Downloads$ ./Status.AppImage 
```
Log:
[Status.log](https://github.com/status-im/status-react/files/2604031/Status.log)

2) If you install Status by replacing current nightly, new application doesn't detect that you already had accounts, so seems that by default it is pointing to another place in the filesystem (for both Linux and Mac Os)

**Comment by vitvly** [comment]
1. Built Linux package locally, it runs multiple instances without issues. Probably Jenkins cached old app skeleton bundle that does not support ubuntu-server newer port parameter format, so it always runs ubuntu-server on port 5000. Made some changes to the build script to clean up the data folder, but there is some Jenkins problem now.
2. Could reproduce the problem. It's because current nightly builds create improper Realm file names, e.g. `default.realmdefault.realm`, `default.realmdefault.realm.management` (notice the duplication, it adds a `default.realm` prefix to all names, while it should have been a parent directory). This is fixed in this PR, so should not be a problem in the future. However, this means an additional folder renaming might be necessary:
```
cd $STATUS_DEFAULT_DIR # e.g. ~/.local/share/Status on linux on ~/Library/Application\ Support/Status on macOS
mv default.realmaccounts default.realm/accounts
mv default.realmdefault.realm default.realm/default.realm
mv default.realmdefault.realm.lock default.realm/default.realm.lock
mv default.realmdefault.realm.management default.realm/default.realm.management
mv default.realmdefault.realm.note default.realm/default.realm.note
```

**Comment by pedropombeiro** [comment]
@siphiuel Maybe we can do the renaming at app startup so we don't bother users with that?

**Comment by churik** [comment]
@siphiuel 
After renaming folders all data (chats) inside my accounts were erased.
Steps:
- install the last build from https://github.com/status-im/status-react/pull/6666#issuecomment-441192333
- rename folders as it is suggested here https://github.com/status-im/status-react/pull/6666#issuecomment-441134773
- launch app
- login with your account

All chats were erased.

**Comment by vitvly** [comment]
@churik renaming should happen after nightly build install, not vice versa. Sorry, wasn't clear.

**Comment by churik** [comment]
@siphiuel 
**Got "We detected a problem with encryption key"**
Steps:
- install latest nightly
- rename folders as it is suggested here https://github.com/status-im/status-react/pull/6666#issuecomment-441134773
- install the last build from https://github.com/status-im/status-react/pull/6666#issuecomment-441271353
- launch app
- login with your account

a prob
I believe this is bacause PR is not rebased.

As we discussed in sync meeting, folders should be renamed automatically.
Can you implement it this PR and rebase it to latest develop?

**Comment by pedropombeiro** [comment]
So we will be asking end users to do the folder renaming themselves?

**Comment by vitvly** [comment]
@PombeirP No, we will rename automatically during startup.

**Comment by chadyj** [comment]
Computers are great at renaming folders but users often forget about this step.

Can we please handle this automatically?  It will make a big difference to users when we save their data folder!

**Comment by pedropombeiro** [comment]
Great, thanks @siphiuel!

**Comment by vitvly** [comment]
Fixed the Linux build issue (thanks @jakubgs !), implemented auto-migration and rebased with latest develop. However, there is some Jenkins issue with iOS build - is it due to some exceeded timeout?

**Comment by vitvly** [comment]
Build successful now. Checked that:
  - multiple instances can start on Linux
  - default.realm directory structure is properly migrated from old to new format

**Comment by vitvly** [comment]
@Churik QA note: if app is run as usual (without any additional shell params as described above) it will attempt to kill ubuntu-server processes upon startup (introduced in https://github.com/status-im/status-react/pull/6761). This means that if a default nightly installation is run after a custom PR with shell params (`STATUS_DATA_DIR` etc.) it will kill the latter's ubuntu-server process. So it is important to launch the PR processes after the nightly.

**Comment by churik** [comment]
@siphiuel I checked behavior on Mac OSx, and everything works as expected. Also on linux multiple instances can be started.

On Linux folders are not renamed automatically after I launched build, so I cannot see accounts, created on previous nightly. 
That's what I got after installing and launching new build:
```
parallels@parallels-vm:~/.local/share/Status$ ls
default.realm               default.realmdefault.realm.lock        ethereum
default.realmaccounts       default.realmdefault.realm.management  keystore
default.realmdefault.realm  default.realmdefault.realm.note        Status.log

```
Expected that folders will be renamed automatically.
Also can you please fix new conflict, fix issue with Linux and update doc with this info (https://github.com/status-im/status-react/pull/6666#issuecomment-442175656) to prevent improper usage?

Also I'm not sure should it be checked on Windows or not.
Thanks!

**Comment by vitvly** [comment]
Updated doc: https://github.com/status-im/status.im/pull/166

**Comment by vitvly** [comment]
@churik thanks for retesting! Fixed the conflict. About renaming - I cannot reproduce this behavior. What I did:
  - downloaded latest nightly and launched it
  - created some account there, and checked that there are folders with wrong names (e.g. `default.realmdefault.realm` in `~/.local/share/Status`
  - quit that instance
  - launched the latest PR build (
https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181127-175738-f203e2-pr.AppImage) and was able to login into the account created in previous steps.
  - also in `~/.local/share/Status/` there were no folders with wrong names, all Realm files have been moved to `default.realm` subfolder.

<blockquote></blockquote>

**Comment by churik** [comment]
Checked:
- installing PR build without parameters - folder names are changed and you can use existing accounts (created in current nightly) (Linux, MacOSx)
- launching two instances of current PR at the same machine - can use (Linux, Mac OSx)
- upgrade to current PR in Windows, IOS as regression. Can login with existing account and create new account.

**Comment by chadyj** [comment]
Nice work everyone! I'm looking forward to testing out more PR's without fear of my db getting nuked! 💥

**Pull request merged**